emmet-core/emmet/core/electronic_structure.py

```diff
@@ -0,0 +1,308 @@
+
+
+class ElectronicStructureSummary(ElectronicStructureBaseData):
+    is_gap_direct: bool = Field(..., description="Whether the band gap is direct.")
+
+    is_metal: bool = Field(..., description="Whether the material is a metal.")
+
+    magnetic_ordering: Union[str, Ordering] = Field(..., description="Magnetic ordering of the calculation.")
+
+
+class BandStructureSummaryData(ElectronicStructureSummary):
+    nbands: float = Field(..., description="Number of bands.")
+
+    equivalent_labels: Dict = Field(..., description="Equivalent k-point labels in other k-path conventions.")
+
+    direct_gap: float = Field(..., description="Direct gap energy in eV.")
+
+
+class DosSummaryData(ElectronicStructureBaseData):
+    spin_polarization: float = Field(None, description="Spin polarization at the fermi level.")
+
+
+class BandstructureData(BaseModel):
+    setyawan_curtarolo: BandStructureSummaryData = Field(
+        None, description="Band structure summary data using the Setyawan-Curtarolo path convention."
+    )
+
+    hinuma: BandStructureSummaryData = Field(
+        None, description="Band structure summary data using the Hinuma et al. path convention."
+    )
+
+    latimer_munro: BandStructureSummaryData = Field(
+        None, description="Band structure summary data using the Latimer-Munro path convention."
+    )
+
+
+class DosData(BaseModel):
+    total: Dict[Union[Spin, str], DosSummaryData] = Field(None, description="Total DOS summary data.")
+
+    elemental: Dict[
+        Element,
+        Dict[
+            Union[Literal["total", "s", "p", "d", "f"], OrbitalType],
+            Dict[Union[Literal["1", "-1"], Spin], DosSummaryData],
+        ],
+    ] = Field(None, description="Band structure summary data using the Hinuma et al. path convention.")
+
+    orbital: Dict[
+        Union[Literal["total", "s", "p", "d", "f"], OrbitalType], Dict[Union[Literal["1", "-1"], Spin], DosSummaryData]
+    ] = Field(None, description="Band structure summary data using the Latimer-Munro path convention.")
+
+    magnetic_ordering: Union[str, Ordering] = Field(None, description="Magnetic ordering of the calculation.")
+
+
+T = TypeVar("T", bound="ElectronicStructureDoc")
+
+
+class ElectronicStructureDoc(PropertyDoc, ElectronicStructureSummary):
+    """
+    Definition for a core Electronic Structure Document
+    """
+
+    bandstructure: BandstructureData = Field(None, description="Band structure data for the material.")
+
+    dos: DosData = Field(None, description="Density of states data for the material.")
+
+    last_updated: datetime = Field(
+        description="Timestamp for when this document was last updated", default_factory=datetime.utcnow
+    )
+
+    @classmethod
+    def from_bsdos(  # type: ignore[override]
+        cls: Type[T],
+        material_id: Union[MPID, int],
+        dos: Dict[Union[MPID, int], CompleteDos],
+        is_gap_direct: bool,
+        is_metal: bool,
```

Comment:
Confused why these are parameters? Can you add arg description to the docstring?

Reply:
These are passed as you can't get `is_gap_direct` or `is_metal` from a DOS object. Since that is the preferred source right now for `band_gap`, `cbm` and `vbm`, I have it manually given in order to populate those at the root level of the doc. There is logic in the builder for which line-mode or static calculation is used.

Yes, I will update the docstring.